dac revocation: Default implementation to check if DAC chain is revoked

shubhamdp opened the pull request:
Fixes #33124
Fixes #33652

#### Change overview
- Impl for parsing revocation set generated using `credentials/generate-revocation-set.py` using jsoncpp.
- Impl for parsing AKID, serial number, issuer name from cert in hex string, and base64 format
- Added an option to chip-tool to take json file as an input revoked set

#### Testing
- Verified commissioning with lighting-app with/without `--dac-revocation-set-path` option.
- When using `--dac-revocation-set-path`, verified with revoked as well as not revoked test dataset.
- Added unit tests

tcarmelveilleux commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -26,10 +26,16 @@
 
+#include <fstream>
+#include <json/json.h>
+#include <string.h>
```
Use `#include <string>` for STL.

tcarmelveilleux commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+
+    for (const auto & revokedSet : jsonData)
+    {
+        if (strncmp(revokedSet["issuer_name"].asCString(), issuerNameBase64Str.data(), issuerNameBase64Str.size()) == 0 &&
+            strncmp(revokedSet["issuer_subject_key_id"].asCString(), akidHexStr.data(), akidHexStr.size()) == 0)
```
Don't use strncmp to do these checks. Construct std::string from .data()/.size()

bzbarsky-apple replied:
💯 

The checks you have here are:

> does `revokedSet["issuer_name"]` start with `issuerNameBase64Str`?

This is not the right check, and we should have tests that would have caught this bug, please.

tcarmelveilleux commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+    VerifyOrReturnError(err == CHIP_NO_ERROR, err);
+    VerifyOrReturnError(outSerialNumberHexStr.size() > serialNumber.size() * 2, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    Encoding::HexFlags flags = Encoding::HexFlags::kUppercaseAndNullTerminate;
+    err = BytesToHex(serialNumber.data(), serialNumber.size(), outSerialNumberHexStr.data(), outSerialNumberHexStr.size(), flags);
+    VerifyOrReturnError(err == CHIP_NO_ERROR, err);
```
Extract hex-ification to a helper, with ByteSpan input and returning an optional<std::string>

tcarmelveilleux reviewed (changes requested):
- [ ] Need to introduce an interface for revocation checking and injection of implementation.
- [ ] Need to avoid JSON and files being added at all to common code outside client-provided implementation.
- [ ] Need `CheckForRevokedDACChain` to be able to push exact errors
- [ ] Need unit tests for the additions

bzbarsky-apple commented on examples/chip-tool/commands/common/CredentialIssuerCommands.h:
```diff
@@ -62,6 +62,15 @@ class CredentialIssuerCommands
+     * @brief
+     *   This function is used to set the path to Device Attestation revocation set JSON file.
+     *
+     * @param[in] path Path to the JSON file containing list of revoked DACs or PAIs.
```
Apparently passing null is allowed?  This needs to be documented, along with what the semantics of passing null are.

bzbarsky-apple commented on examples/chip-tool/commands/common/CredentialIssuerCommands.h:
```diff
@@ -62,6 +62,15 @@ class CredentialIssuerCommands
+     * @param[in] path Path to the JSON file containing list of revoked DACs or PAIs.
+     *                 It can be generated using credentials/generate-revocation-set.py script
+     */
+    virtual void SetupDeviceAttestationRevocationSetPath(const char * path) = 0;
```
```suggestion
    virtual void SetDeviceAttestationRevocationSetPath(const char * path) = 0;
```

bzbarsky-apple commented on examples/chip-tool/commands/example/ExampleCredentialIssuerCommands.h:
```diff
@@ -44,6 +44,17 @@ class ExampleCredentialIssuerCommands : public CredentialIssuerCommands
+    {
+        if (path)
+        {
+            // As we know that we are using DefaultDACVerifier, we can downcast from
```
If we know that, why is `mDacVerifier` not just of type `DefaultDACVerifier`?  Either it can be (and then this cast is not needed), or it can't be, and then this cast is not valid.  I think it's the latter, since nothing says `GetDefaultDACVerifier` in fact returns an instance of `DefaultDACVerifier` right now, and its implementation could easily change to return something else in some cases.

Either way, there should be no cast here.

bzbarsky-apple commented on examples/chip-tool/commands/example/ExampleCredentialIssuerCommands.h:
```diff
@@ -44,6 +44,17 @@ class ExampleCredentialIssuerCommands : public CredentialIssuerCommands
+
+    void SetupDeviceAttestationRevocationSetPath(const char * path) override
+    {
+        if (path)
```
OK, so if path is null then we keep using whatever the previous path used to be?  Again, this all needs to be clearly documented in the API, and then we can have a discussion about what the right behavior is.

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+//  }
+// ]
+//
+bool DefaultDACVerifier::IsEntryExistsInRevocationSet(const CharSpan & akidHexStr, const CharSpan & issuerNameBase64Str,
```
```suggestion
bool DefaultDACVerifier::IsEntryInRevocationSet(const CharSpan & akidHexStr, const CharSpan & issuerNameBase64Str,
```

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+
+    if (!parsingSuccessful)
+    {
+        return false;
```
Please log the errors; otherwise this is a completely inscrutable failure.

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+        if (strncmp(revokedSet["issuer_name"].asCString(), issuerNameBase64Str.data(), issuerNameBase64Str.size()) == 0 &&
+            strncmp(revokedSet["issuer_subject_key_id"].asCString(), akidHexStr.data(), akidHexStr.size()) == 0)
+        {
+            for (const auto & revokedSerialNumber : revokedSet["revoked_serial_numbers"])
```
Consider reversing the sense of the issuer name/subject checks, doing `continue` if either of those does not match, and outdenting.

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+    MutableByteSpan akid(akidBuf);
+
+    CHIP_ERROR err = ExtractAKIDFromX509Cert(certDer, akid);
+    VerifyOrReturnError(err == CHIP_NO_ERROR, err);
```
```suggestion
    ReturnErrorOnFailure(err);
```

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+    VerifyOrReturnError(err == CHIP_NO_ERROR, err);
+    VerifyOrReturnError(outAKIDHexStr.size() > akid.size() * 2, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    Encoding::HexFlags flags = Encoding::HexFlags::kUppercaseAndNullTerminate;
```
Why are we null-terminating?  We shouldn't need to null-terminate if we are writing into a span, which knows its own length.  And then we can make the length check `>=`, not `>`.

That said, doesn't `BytesToHex` do its own size check?  Why are we doing a separate one?

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+
+    Encoding::HexFlags flags = Encoding::HexFlags::kUppercaseAndNullTerminate;
+    err                      = BytesToHex(akid.data(), akid.size(), outAKIDHexStr.data(), outAKIDHexStr.size(), flags);
+    VerifyOrReturnError(err == CHIP_NO_ERROR, err);
```
```suggestion
    ReturnErrorOnFailure(err);
```

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+    err                      = BytesToHex(akid.data(), akid.size(), outAKIDHexStr.data(), outAKIDHexStr.size(), flags);
+    VerifyOrReturnError(err == CHIP_NO_ERROR, err);
+
+    outAKIDHexStr.reduce_size(strlen(outAKIDHexStr.data()));
```
This makes no sense, but does explain why we were null-terminating...

The size must be `akid.size() * 2`, no?

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+    MutableByteSpan serialNumber(serialNumberBuf);
+
+    CHIP_ERROR err = ExtractSerialNumberFromX509Cert(certDer, serialNumber);
+    VerifyOrReturnError(err == CHIP_NO_ERROR, err);
```
A bunch of the GetAKIDHexStr comments apply here.

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+    MutableByteSpan issuer(issuerBuf);
+
+    CHIP_ERROR err = ExtractIssuerFromX509Cert(certDer, issuer);
+    VerifyOrReturnError(CHIP_NO_ERROR == err, err);
```
See comments above.

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+
+    CHIP_ERROR err = ExtractIssuerFromX509Cert(certDer, issuer);
+    VerifyOrReturnError(CHIP_NO_ERROR == err, err);
+    VerifyOrReturnError(outIssuerNameBase64String.size() > BASE64_ENCODED_LEN(issuer.size()), CHIP_ERROR_BUFFER_TOO_SMALL);
```
Why is this a `>`, not a `>=` check?  I would expect passing a buffer of size `BASE64_ENCODED_LEN(dataToEncode)` to work....

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+    MutableCharSpan akid(akidHexStrBuffer);
+
+    CHIP_ERROR err = GetIssuerNameBase64Str(certDer, issuerName);
+    VerifyOrReturnValue(err == CHIP_NO_ERROR, false);
```
ReturnErrorOnFailure various places here.

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
+        }
+        if (IsCertificateRevoked(info.paiDerBuffer))
+        {
+            attestationError = AttestationVerificationResult::kPaiRevoked;
```
So if both are revoked we consider the PAI being revoked more important than the DAC being revoked, in terms of the errors that get reported?

shubhamdp replied:
Added another error code `kPaiAndDacRevoked` and using that one.

bzbarsky-apple commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.cpp:
```diff
@@ -607,12 +613,147 @@ CHIP_ERROR DefaultDACVerifier::VerifyNodeOperationalCSRInformation(const ByteSpa
     return CHIP_NO_ERROR;
 }
 
+// This method parses the below JSON Scheme
```
So right now, you can provide a trust store to a DefaultDACVerifier and it will work with that via API, agnostic to how the data is actually stored or gotten.

We should probably do something similar here; there are existing deployments of DefaultDACVerifier that cannot in fact create a file on disk to point it to, so that detail should be hidden behind an API boundary, with some TestSomething class that implements the relevant API and has the JSON thing as an implementation.

The API to that thing should ideally be async, just like the API to the DefaultDACVerifier itself.

shubhamdp commented:
@tcarmelveilleux I have address the comments, PTAL for more details below

> * [x]  Need to introduce an interface for revocation checking and injection of implementation.

Added the `DeviceAttestationRevocationDelegate` and methods to inject the implementation into DeviceAttestationVerifier.
> * [x]  Need to avoid JSON and files being added at all to common code outside client-provided implementation.

Not using JSON impl in the DefaultDeviceAttestionVerifier and added TestDACRevocationDelegateImpl.
> * [x]  Need `CheckForRevokedDACChain` to be able to push exact errors

Added another error code if PAI and DAC both are revoked.
> * [x]  Need unit tests for the additions

Added unit tests

shubhamdp commented:
@bzbarsky-apple I've address the review-comments PTAL!

bzbarsky-apple commented on examples/chip-tool/commands/common/CredentialIssuerCommands.h:
```diff
@@ -62,6 +62,17 @@ class CredentialIssuerCommands
 
+    /**
+     * @brief
+     *   This function is used to set the path to Device Attestation revocation set JSON file.
```
```suggestion
     *   This function is used to set the path to the Device Attestation revocation set JSON file.
```

bzbarsky-apple commented on src/credentials/attestation_verifier/DeviceAttestationVerifier.h:
```diff
@@ -79,6 +79,8 @@ enum class AttestationVerificationResult : uint16_t
 
     kInternalError = 900,
 
+    kPaiAndDacRevoked = 1000,
```
I'd expect this to be up near the other Pai/Dac bits, with a similar value.  Not sure whether to put it with the kDac or kPai constants; which of those is the more immediate problem if both are revoked?

shubhamdp replied:
Moved closer to 'kPai` bits, as the name starts with 'kPai` and  revoked PAI is a bigger concern, revoked DAC.

bzbarsky-apple commented on src/credentials/attestation_verifier/DeviceAttestationVerifier.h:
```diff
@@ -409,13 +413,39 @@ class DeviceAttestationVerifier
     // always supported development keys.
     bool mEnableCdTestKeySupport = true;
+
+    DeviceAttestationRevocationDelegate * mRevocationDelegate = nullptr;
```
Why is this on the interface here?  Shouldn't it be on whatever class uses DeviceAttestationRevocationDelegate to implement CheckForRevokedDACChain?

I guess the issue is that some consumer (e.g. chip-tool?) just does GetDefaultDACVerifier() and then wants to set this delegate on it. But this part doesn't make sense for other verifier impls....

What should probably happen is that `GetDefaultDACVerifier` grows either an optional argument or a two-argument overload or something.

shubhamdp replied:
Even I didn't want it here, but wasn't sure where to put it. Thanks for the suggestion. Moved to `DefaultDACVerifier`. and adjusted the bits in chip-tool to set it up accordingly. PTAL

bzbarsky-apple commented on examples/chip-tool/commands/common/CHIPCommand.cpp:
```diff
@@ -450,7 +453,14 @@ CHIP_ERROR CHIPCommand::InitializeCommissioner(CommissionerIdentity & identity,
-    ReturnLogErrorOnFailure(mCredIssuerCmds->SetupDeviceAttestation(commissionerParams, sTrustStore));
+    if (mDacRevocationSetPath.HasValue())
+    {
+        static chip::Credentials::TestDACRevocationDelegateImpl testDacRevocationDelegate;
```
So in interactive mode, if `mDacRevocationSetPath` has no value this will just keep using the revocation set from last time?  Trying to understand the intended behavior here.

shubhamdp replied:
Yes that is the intended behaviour here, keeping it similar to trust store path. Also, factored out this code into similar method as trust store path.

should we honour the argument provided in interactive mode? If yes, I'll do it in follow up.

bzbarsky-apple replied:
This is quite different from the trust store path behavior.  The behavior for trust store path is:

1. The first time a command is executed in interactive mode with a trust store path argument, create a FileAttestationTrustStore for that path.
2. Any time after that a command is executed with a trust store path argument, use that FileAttestationTrustStore (with the old path, ignoring the new value).
3. Any command executed without a trust store path argument uses `Credentials::GetTestAttestationTrustStore()`. 
4. Some complications around the env var bits, but let's ignore those.

The behavior that is being implemented here, on the other hand is:

1. The first time a command is executed in interactive mode with a trust store path argument, create a TestDACRevocationDelegateImpl for that path.
2. Any time after that a command is executed, use that same TestDACRevocationDelegateImpl no matter what the revocation path for the command was or whether one was present.

Neither behavior is terribly sane, honestly, but they are also different and I am not sure whether the behavior differences are intentional or not.

shubhamdp replied:
Thanks for the explanation. Yes, they are not exactly similar. May be once we have default revocation set built up, I can align the behaviour to attestation trust store. I have added the comment in the code, about how it behaves. Also raised #34434

tcarmelveilleux commented on src/credentials/attestation_verifier/DeviceAttestationVerifier.h:
```diff
@@ -47,6 +47,7 @@ enum class AttestationVerificationResult : uint16_t
     kPaiVendorIdMismatch  = 205,
     kPaiAuthorityNotFound = 206,
     kPaiMissing           = 207,
+    kPaiAndDacRevoked     = 208,
```
Have a different code for PAI revoked, and DAC revoked, in the correct ranges (200 for PAI, 300 for DAC)

bzbarsky-apple replied:
Those already exist.  This is the code for "they are both revoked".

tcarmelveilleux commented on src/credentials/attestation_verifier/DeviceAttestationVerifier.h:
```diff
@@ -259,6 +260,8 @@ class ArrayAttestationTrustStore : public AttestationTrustStore
     const size_t mNumCerts;
 };
 
+class DeviceAttestationRevocationDelegate;
```
This would not be needed if DeviceAttestationRevocationDelegate was a subclass of DeviceAttestationVerifier

shubhamdp replied:
Logged an issue to make it a subclass, will do it in follow-up https://github.com/project-chip/connectedhomeip/issues/34495

tcarmelveilleux commented on src/credentials/attestation_verifier/TestDACRevocationDelegateImpl.cpp:
```diff
@@ -0,0 +1,204 @@
+//  {
+//    "type": "revocation_set",
+//    "issuer_subject_key_id": "63540E47F64B1C38D13884A462D16C195D8FFB3C",
+//    "issuer_name": "MD0xJTAjBgNVBAMMHE1hdHRlciBEZXYgUEFJIDB4RkZGMSBubyBQSUQxFDASBgorBgEEAYKifAIBDARGRkYx",
```
What constitutes the issuer name? Which bytes?

bzbarsky-apple commented on examples/chip-tool/commands/common/CHIPCommand.h:
```diff
@@ -222,11 +224,16 @@ class CHIPCommand : public Command
     // can spin up commissioners as needed.
     static const chip::Credentials::AttestationTrustStore * sTrustStore;
 
+    // Cached DAC revocation delegate, this can be set using "--dac-revocation-set-path" argument
+    // Once set this will be used by all commands.
```
This information should be in the AddArgument help message, no?  People should not have to read the source to find out how arguments behave.

shubhamdp replied:
Added it to the argument help!

dhrishi commented:
@tcarmelveilleux @andy31415 Can you please review and approve this if it looks okay? We are hoping that this gets merged before the 1.4.TE2 deadline

tcarmelveilleux commented on src/credentials/BUILD.gn:
```diff
@@ -157,6 +158,8 @@ static_library("default_attestation_verifier") {
     "attestation_verifier/DefaultDeviceAttestationVerifier.cpp",
     "attestation_verifier/DefaultDeviceAttestationVerifier.h",
     "attestation_verifier/DeviceAttestationDelegate.h",
+    "attestation_verifier/TestDACRevocationDelegateImpl.cpp",
+    "attestation_verifier/TestDACRevocationDelegateImpl.h",
```
Remove these files from here. and the `jsoncpp_root` dependency in default_attestation_verifier

shubhamdp replied:
Thanks for pointing it out, removed. Thats why tests were passing!

tcarmelveilleux commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.h:
```diff
@@ -57,7 +57,11 @@ class CsaCdKeysTrustStore : public WellKnownKeysTrustStore
 public:
-    DefaultDACVerifier(const AttestationTrustStore * paaRootStore) : mAttestationTrustStore(paaRootStore) {}
+    DefaultDACVerifier(const AttestationTrustStore * paaRootStore,
+                       DeviceAttestationRevocationDelegate * revocationDelegate = nullptr) :
```
Instead of having a default parameter, just add a second constructor with both arguments. This will keep backwards compatibility.

shubhamdp replied:
Added one more constructor.

tcarmelveilleux commented on src/credentials/attestation_verifier/DefaultDeviceAttestationVerifier.h:
```diff
@@ -84,6 +88,7 @@ class DefaultDACVerifier : public DeviceAttestationVerifier
 
     CsaCdKeysTrustStore mCdKeysTrustStore;
     const AttestationTrustStore * mAttestationTrustStore;
+    DeviceAttestationRevocationDelegate * mRevocationDelegate = nullptr;
```
Suggest having a setter for the revocation delegate.

shubhamdp replied:
Added the setter API

tcarmelveilleux commented on src/credentials/attestation_verifier/TestDACRevocationDelegateImpl.cpp:
```diff
@@ -0,0 +1,205 @@
+    ReturnErrorOnFailure(ExtractIssuerFromX509Cert(certDer, issuer));
+    VerifyOrReturnError(outIssuerNameBase64String.size() >= BASE64_ENCODED_LEN(issuer.size()), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    uint32_t encodedLen = Base64Encode32(issuer.data(), static_cast<uint32_t>(issuer.size()), outIssuerNameBase64String.data());
```
This can be simply `Base64Encode`.  I don't know what the `32` suffix is about.

shubhamdp replied:
32 suffix returns the 32 bit value. `Base64Encode` returns 16 bit value. Using `Base64Encode`.

tcarmelveilleux commented on src/credentials/attestation_verifier/TestDACRevocationDelegateImpl.cpp:
```diff
@@ -0,0 +1,205 @@
+    MutableCharSpan serialNumber(serialNumberHexStrBuffer);
+    MutableCharSpan akid(akidHexStrBuffer);
+
+    VerifyOrReturnValue(CHIP_NO_ERROR == GetIssuerNameBase64Str(certDer, issuerName), false);
+    ChipLogDetail(NotSpecified, "Issuer: %.*s", static_cast<int>(issuerName.size()), issuerName.data());
+
+    VerifyOrReturnValue(CHIP_NO_ERROR == GetSerialNumberHexStr(certDer, serialNumber), false);
+    ChipLogDetail(NotSpecified, "Serial Number: %.*s", static_cast<int>(serialNumber.size()), serialNumber.data());
+
+    VerifyOrReturnValue(CHIP_NO_ERROR == GetAKIDHexStr(certDer, akid), false);
+    ChipLogDetail(NotSpecified, "AKID: %.*s", static_cast<int>(akid.size()), akid.data());
```
Add a TODO (and create associated issue) to cross-validate the CRLSignerCertificate and CRLSignerDelegator per spec:

```
. If a matching `RevocationSet` is found, determined if `serialNumber` is a member of the set.
    .. If the entity type whose revocation status is being verified is a PAI, then the subject PAI's issuer (a PAA) SHALL fulfill one of the following two cases, otherwise return immediately:
      ... The Subject and Subject Key of the PAI certificate's issuer matches exactly the CRLSignerCertificate's subject (i.e. the CRLSignerCertificate is a PAA).
      ... The Subject and Subject Key of the PAI certificate's issuer matches exactly the PAA which is the issuer of the CRLSignerCertificate (i.e. the CRLSignerCertificate is a CRL signer delegated by a PAA).
    .. If the entity type whose revocation status is being verified is a DAC, then:
      ... If the `CRLSignerDelegator` is present, then the subject DAC's issuer (a PAI) SHALL match the `CRLSignerDelegator` in both Subject Key and Subject, otherwise return immediately.
      ... If the `CRLSignerDelegator` is absent, then the subject DAC's issuer (a PAI) SHALL match the `CRLSignerCertificate` in both Subject Key and Subject, otherwise return immediately.
```

Those checks are not currently do, and MUST be done later once this is supported in revocation set construction.

shubhamdp replied:
Added a TODO and an issue: #34587

tcarmelveilleux commented on src/credentials/attestation_verifier/TestDACRevocationDelegateImpl.cpp:
```diff
@@ -0,0 +1,205 @@
+{
+    AttestationVerificationResult attestationError = AttestationVerificationResult::kSuccess;
+
+    if (mDeviceAttestationRevocationSetPath != nullptr)
```
Suggest an early exit with call of the callback if `mDeviceAttestationRevocationSetPath == nullptr`, to de-nest this next part.

shubhamdp replied:
Since we have to make a call at the end with attestation result, that why it was like this. Inverted the check and returning early.

tcarmelveilleux commented on src/credentials/attestation_verifier/TestDACRevocationDelegateImpl.cpp:
```diff
@@ -0,0 +1,205 @@
+
+    if (mDeviceAttestationRevocationSetPath != nullptr)
+    {
+        ChipLogDetail(NotSpecified, "Checking for revoked DAC in %s", mDeviceAttestationRevocationSetPath);
```
Suggeste using std::string for the mDeviceAttestationRevocationSetPath

shubhamdp replied:
Using earlier suggestion, `std::string_view`

tcarmelveilleux commented on src/credentials/tests/TestDeviceAttestationCredentials.cpp:
The test needs to now have a dependency on `:test_dac_revocation_delegate` target.

shubhamdp replied:
Added!

tcarmelveilleux commented on examples/chip-tool/commands/common/CredentialIssuerCommands.h:
```diff
@@ -57,10 +57,14 @@ class CredentialIssuerCommands
-    virtual CHIP_ERROR SetupDeviceAttestation(chip::Controller::SetupParams & setupParams,
-                                              const chip::Credentials::AttestationTrustStore * trustStore) = 0;
+    virtual CHIP_ERROR
+    SetupDeviceAttestation(chip::Controller::SetupParams & setupParams, const chip::Credentials::AttestationTrustStore * trustStore,
```
Please do not use default params for this, just require the value to always be passed.

shubhamdp replied:
not using the default value.

shubhamdp commented on src/credentials/attestation_verifier/TestDACRevocationDelegateImpl.cpp:
```diff
@@ -0,0 +1,205 @@
+}
+} // anonymous namespace
+
+CHIP_ERROR TestDACRevocationDelegateImpl::SetDeviceAttestationRevocationSetPath(const char * path)
```
Using `std::string_view` for this.

shubhamdp commented on src/credentials/tests/TestDeviceAttestationCredentials.cpp:
```diff
@@ -413,3 +416,158 @@ TEST_F(TestDeviceAttestationCredentials, TestAttestationTrustStore)
+    file.open(fileName, std::ofstream::out | std::ofstream::trunc);
+    file << jsonData;
+    file.close();
+}
```
Added TODO for now, will fix in follow-up #34588

tcarmelveilleux commented:
Broke python tests on HEAD
